### PR TITLE
Implement REPLACE operation for EnvoyFilter patch

### PR DESCRIPTION
**Pull request by zhaohuabing**
Add a "REPLACE" operation for EnvoyFilter patch. This operation only supports Network filter and HTTP filter.

This is the implementation of issue #27425

Signed-off-by: huabing zhao <zhaohuabing@gmail.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[*] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go [hunk 1]
```diff
@@ -437,6 +453,26 @@ func doHTTPFilterListOperation(patchContext networking.EnvoyFilter_PatchContext,
...
+			// find the matching filter first
+			replacePosition := -1
+			for i := 0; i < len(hcm.HttpFilters); i++ {
+				if httpFilterMatch(hcm.HttpFilters[i], cp) {
+					replacePosition = i
+					break
+				}
+			}
+
+			if replacePosition == -1 {
```
Comment (mandarjog) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go:470]: Best to add debug log here.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go [hunk 2]
```diff
@@ -465,9 +518,54 @@ func TestApplyListenerPatches(t *testing.T) {
...
+						{Name: "filter2"},
+					},
+				},
+			},
+		},
 		{
 			Name: "another-listener",
 		},
+		{
+			Name: "listener-to-be-replaced",
```
Comment (mandarjog) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go:545]: Nit: change name to indicate http filter

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go [hunk 3]
```diff
@@ -465,9 +518,54 @@ func TestApplyListenerPatches(t *testing.T) {
...
+			FilterChains: []*listener.FilterChain{
+				{
+					FilterChainMatch: &listener.FilterChainMatch{
+						DestinationPort: &wrappers.UInt32Value{
+							Value: 80,
+						},
+					},
```
Comment (mandarjog) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch_test.go:552]: Add negative test, when to-be replaced filter is not found.

**Review by mandarjog — COMMENTED** [review]
Thanks for doing this.

**Review by hzxuzhonghu — APPROVED** [review]

**Review by mandarjog — APPROVED** [review]
Lgtm

File: releasenotes/notes/27425.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 27425
+releaseNotes:
+  - |
+    **Added** "REPLACE" operation for EnvoyFilter. "REPLACE" opration can replace contents of a named filter with new contents. It is only valid for HTTP_FILTER and NETWORK_FILTER.
```
Comment (howardjohn) [inline releasenotes/notes/27425.yaml:8]:
```suggestion
    **Added** "REPLACE" operation for EnvoyFilter. "REPLACE" operation can replace contents of a named filter with new contents. It is only valid for HTTP_FILTER and NETWORK_FILTER. 
```

Comment (zhaohuabing) [reply releasenotes/notes/27425.yaml:8]: Sorry for the typo, how could I fix it? This has already been merged.

**Comment by howardjohn** [comment]
You can send another PR I guess. Sorry didn't see it had been merged.
Thanks for the PR!

**Pull request merged**